src/subsplit_dag.cpp

```diff
@@ -578,6 +608,264 @@ void SubsplitDAG::ReversePostorderIndexTraversal(
+  edge_reindexer = BuildEdgeReindexer(prev_edge_count);
+  // Update new_node_ids and new_edge_idxes according to the reindexers.
+  for (size_t &id : new_node_ids) {
+    id = node_reindexer.at(id);
+  }
+  for (size_t &idx : new_edge_idxes) {
+    idx = edge_reindexer.at(idx);
+  }
+  // Update fields in the Subsplit DAG according to the reindexers.
+  UpdateNodeIds(node_reindexer);
+  UpdateEdgeIdxs(edge_reindexer);
+  // Recount topologies to update `topology_count_below_`.
+  CountTopologies();
+  return {new_node_ids, new_edge_idxes, node_reindexer, edge_reindexer};
+}
+
+SizeVector SubsplitDAG::BuildNodeReindexer(const size_t prev_node_count) {
+  SizeVector node_reindexer = Reindexer::IdentityReindexer(NodeCount());
+  size_t running_traversal_idx = taxon_count_;
+  size_t dag_root_node_id = prev_node_count - 1;
+  // Build node_reindexer by traversing entire DAG and assigning new ids after child ids
+  // are assigned.
+  DepthFirstWithAction(
+      {dag_root_node_id},
+      SubsplitDAGTraversalAction(
+          // BeforeNode
+          [this](size_t node_id) {},
+          // AfterNode
+          [this, &node_reindexer, &running_traversal_idx](size_t node_id) {
+            node_reindexer.at(node_id) = running_traversal_idx;
+            running_traversal_idx++;
+          },
+          // BeforeNodeClade
+          [this](size_t node_id, bool rotated) {},
+          // VisitEdge
+          [this](size_t node_id, size_t child_id, bool rotated) {}));
+  return node_reindexer;
+}
+
+SizeVector SubsplitDAG::BuildEdgeReindexer(const size_t prev_edge_count) {
+  SizeVector edge_reindexer =
+      Reindexer::IdentityReindexer(GPCSPCountWithFakeSubsplits());
+  // Only edges from an existing parent node to a new child node need to be reindexed.
+  // See SubsplitDAG::AddNodePair().
+  for (size_t edge_idx = prev_edge_count; edge_idx < GPCSPCountWithFakeSubsplits();
+       edge_idx++) {
+    size_t parent_id;
+    size_t child_id;
+    // Find the nodes that the new edge connects.
+    for (const auto &[node_pair, idx] : dag_edges_) {
+      if (idx == edge_idx) {
+        parent_id = node_pair.first;
+        child_id = node_pair.second;
+        break;
+      }
+    }
+    Bitset parent_subsplit = GetDAGNode(parent_id)->GetBitset();
+    Bitset child_subsplit = GetDAGNode(child_id)->GetBitset();
+    const auto idx_range = GetEdgeRange(
+        parent_subsplit, child_subsplit.SubsplitIsRotatedChildOf(parent_subsplit));
+    // New edge is added to the end of the range.
+    const size_t new_idx(idx_range.second);
+    edge_reindexer.at(edge_idx) = edge_reindexer.at(new_idx);
+    for (size_t idx = new_idx; idx < edge_idx; idx++) {
+      // Every idx after the new edge's position is shifted by 1.
+      if (edge_reindexer.at(idx) >= new_idx) {
+        edge_reindexer.at(idx) = edge_reindexer.at(idx) + 1;
+      }
+    }
+  }
+  return edge_reindexer;
+}
+
+void SubsplitDAG::UpdateNodeIds(SizeVector node_reindexer) {
```

Comment:
Could be const ref, right?

Reply:
Also, I think that a better name for this would be "ApplyNodeReindexer". What do you think?